File: backend/tcgplayer.py
import re
import httpx
from typing import List, Dict, Any, Optional
from config.settings import TCGPLAYER_SEARCH_API, HEADERS

def extract_card_number(raw_input: str) -> Dict[str, str]:
    """
    Extracts the core card number identifier (e.g. 'OP13-120' from 'OP13-120 SEC')
    and any specified rarity / variant keywords.
    """
    text = raw_input.strip()
    
    # 1. Match card numbers like OP13-120, ST30-001, CORI-EN062, 081/088, 4/102
    pattern = r'([A-Za-z0-9]+[-/][0-9A-Za-z]+|\b\d{1,3}/\d{1,3}\b)'
    match = re.search(pattern, text)
    
    card_number = match.group(1) if match else text

    # Remove common rarity keywords if appended
    rarity_keywords = ["SEC", "SAR", "SR", "SP", "L", "R", "UC", "C", "PARALLEL", "ALT", "HOLO", "FOIL", "PROMO"]
    tokens = text.upper().split()
    detected_rarities = [t for t in tokens if t in rarity_keywords]

    return {
        "raw": text,
        "extracted_number": card_number,
        "rarities": detected_rarities
    }

def search_card_by_number(card_input: str) -> List[Dict[str, Any]]:
    clean_input = card_input.strip()
    if not clean_input:
        return []

    parsed = extract_card_number(clean_input)
    target_num = parsed["extracted_number"]

    # Strategy 1: Exact Number term filter (e.g. number: ["OP13-120"])
    payload_term = {
        "algorithm": "free_text_search",
        "from": 0,
        "size": 30,
        "filters": {
            "term": {
                "number": [target_num]
            }
        },
        "query": "",
        "context": {"shippingCountry": "US", "cart": {}}
    }

    raw_results = []
    try:
        with httpx.Client(timeout=15.0) as client:
            res = client.post(TCGPLAYER_SEARCH_API, json=payload_term, headers=HEADERS)
            if res.status_code == 200:
                raw_results = res.json().get("results", [{}])[0].get("results", [])

            # Strategy 2: Free text search using extracted number if term filter returned empty
            if not raw_results:
                payload_extracted_query = {
                    "algorithm": "free_text_search",
                    "from": 0,
                    "size": 30,
                    "filters": {"term": {}},
                    "query": target_num,
                    "context": {"shippingCountry": "US", "cart": {}}
                }
                res_q = client.post(TCGPLAYER_SEARCH_API, json=payload_extracted_query, headers=HEADERS)
                if res_q.status_code == 200:
                    raw_results = res_q.json().get("results", [{}])[0].get("results", [])

            # Strategy 3: Free text search using raw input
            if not raw_results:
                payload_raw_query = {
                    "algorithm": "free_text_search",
                    "from": 0,
                    "size": 30,
                    "filters": {"term": {}},
                    "query": clean_input,
                    "context": {"shippingCountry": "US", "cart": {}}
                }
                res_raw = client.post(TCGPLAYER_SEARCH_API, json=payload_raw_query, headers=HEADERS)
                if res_raw.status_code == 200:
                    raw_results = res_raw.json().get("results", [{}])[0].get("results", [])
    except Exception as e:
        print(f"Error fetching TCGPlayer data for '{card_input}': {e}")
        return []

    # Filter and parse results
    parsed_variants = []
    seen_ids = set()

    for item in raw_results:
        product_id = item.get("productId")
        if not product_id or product_id in seen_ids:
            continue

        product_id = int(product_id)
        product_name = item.get("productName", "Unknown Card")
        set_name = item.get("setName", "Unknown Set")
        market_price = item.get("marketPrice")
        custom = item.get("customAttributes", {})
        number = custom.get("number", target_num)
        rarity = custom.get("rarity", "")

        # Check if this result matches our target card number (exact or partial match)
        num_clean = str(number).strip().upper()
        target_clean = target_num.strip().upper()

        is_exact_num_match = (num_clean == target_clean or target_clean in num_clean)

        image_url = f"https://tcgplayer-cdn.tcgplayer.com/product/{product_id}_200w.jpg"
        product_url = f"https://www.tcgplayer.com/product/{product_id}"

        seen_ids.add(product_id)
        parsed_variants.append({
            "productId": product_id,
            "productName": product_name,
            "setName": set_name,
            "number": number,
            "rarity": rarity,
            "marketPriceUSD": float(market_price) if market_price is not None else None,
            "imageUrl": image_url,
            "productUrl": product_url,
            "isExactMatch": is_exact_num_match
        })

    # Sort results: exact card number matches first, then by market price
    parsed_variants.sort(key=lambda x: (not x["isExactMatch"], x["marketPriceUSD"] is None))

    return parsed_variants

File: frontend/src/services/tcgplayer.js
// Client-side TCGPlayer search & latest sales engine for standalone GitHub Pages deployment

const TCGPLAYER_SEARCH_API = "https://mp-search-api.tcgplayer.com/v1/search/request";

export function extractCardNumber(rawInput) {
  const text = rawInput.trim();
  const pattern = /([A-Za-z0-9]+[-/][0-9A-Za-z]+|\b\d{1,3}\/\d{1,3}\b)/;
  const match = text.match(pattern);
  const cardNumber = match ? match[1] : text;

  return {
    raw: text,
    extractedNumber: cardNumber
  };
}

export async function fetchRecentSalesClient(productId) {
  const url = `https://mp-search-api.tcgplayer.com/v1/product/${productId}/latestsales`;
  try {
    const res = await fetch(url, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Accept": "application/json"
      },
      body: JSON.stringify({ limit: 10 })
    });

    if (res.ok) {
      const data = await res.json();
      const sales = data.data || [];
      const prices = sales
        .map(s => s.purchasePrice)
        .filter(p => p !== null && p !== undefined)
        .map(p => parseFloat(p));

      const recent3 = prices.slice(0, 3);
      const avg3 = recent3.length > 0 
        ? Math.round((recent3.reduce((a, b) => a + b, 0) / recent3.length) * 100) / 100 
        : null;

      return { recentSalesUSD: recent3, averageRecentSalesUSD: avg3 };
    }
  } catch (err) {
    console.warn(`Client-side sales fetch fallback for product ${productId}:`, err);
  }

  return { recentSalesUSD: [], averageRecentSalesUSD: null };
}

export async function searchCardClientSide(cardInput, onePieceOnly = true) {
  const cleanInput = cardInput.strip ? cardInput.strip() : cardInput.trim();
  if (!cleanInput) return [];

  const { extractedNumber: targetNum } = extractCardNumber(cleanInput);

  const categoryFilter = onePieceOnly ? { productCategoryName: ["One Piece Card Game"] } : {};

  const payloadTerm = {
    algorithm: "free_text_search",
    from: 0,
    size: 30,
    filters: {
      term: {
        number: [targetNum],
        ...categoryFilter
      }
    },
    query: "",
    context: { shippingCountry: "US", cart: {} }
  };

  let rawResults = [];
  try {
    const res = await fetch(TCGPLAYER_SEARCH_API, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "Accept": "application/json"
      },
      body: JSON.stringify(payloadTerm)
    });

    if (res.ok) {
      const data = await res.json();
      rawResults = data.results?.[0]?.results || [];
    }

    if (rawResults.length === 0) {
      const payloadExtracted = {
        algorithm: "free_text_search",
        from: 0,
        size: 30,
        filters: { term: categoryFilter },
        query: targetNum,
        context: { shippingCountry: "US", cart: {} }
      };

      const resEx = await fetch(TCGPLAYER_SEARCH_API, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          "Accept": "application/json"
        },
        body: JSON.stringify(payloadExtracted)
      });

      if (resEx.ok) {
        const dataEx = await resEx.json();
        rawResults = dataEx.results?.[0]?.results || [];
      }
    }

    if (rawResults.length === 0 && onePieceOnly) {
      return searchCardClientSide(cardInput, false);
    }
  } catch (err) {
    console.error(`Client-side fetch error for ${cardInput}:`, err);
    return [];
  }

  const parsedVariants = [];
  const seenIds = new Set();

  for (const item of rawResults) {
    const productId = item.productId;
    if (!productId || seenIds.has(productId)) continue;

    seenIds.add(productId);
    const pId = parseInt(productId);
    const productName = item.productName || "Unknown Card";
    const setName = item.setName || "Unknown Set";
    const marketPrice = item.marketPrice !== undefined && item.marketPrice !== null ? parseFloat(item.marketPrice) : null;
    const custom = item.customAttributes || {};
    const number = custom.number || targetNum;
    const rarity = custom.rarity || "";

    const numClean = String(number).trim().toUpperCase();
    const targetClean = targetNum.trim().toUpperCase();
    const isExactNumMatch = numClean === targetClean || numClean.includes(targetClean);

    const imageUrl = `https://tcgplayer-cdn.tcgplayer.com/product/${pId}_200w.jpg`;
    const productUrl = `https://www.tcgplayer.com/product/${pId}`;

    const { recentSalesUSD, averageRecentSalesUSD } = await fetchRecentSalesClient(pId);

    const finalAvgUSD = averageRecentSalesUSD !== null ? averageRecentSalesUSD : marketPrice;

    parsedVariants.push({
      productId: pId,
      productName,
      setName,
      number,
      rarity,
      marketPriceUSD: marketPrice,
      recentSalesUSD,
      averageRecentSalesUSD: finalAvgUSD,
      imageUrl,
      productUrl,
      isExactMatch: isExactNumMatch
    });
  }

  parsedVariants.sort((a, b) => {
    if (a.isExactMatch !== b.isExactMatch) return a.isExactMatch ? -1 : 1;
    return (a.marketPriceUSD === null) - (b.marketPriceUSD === null);
  });

  return parsedVariants;
}
